docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+   the link:../../tests/API/XCCDF/Makefile.am[tests/API/XCCDF/Makefile.am].
+   When adding a new entry please keep in mind that all entries are sorted in
+   the alphabetical order.
+. Create `Makefile.am` inside the `tests/API/XCCDF/foo` directory with the
+   following contents:
+[[app-listing]]
+[source,bash]
+----
+DISTCLEANFILES = *.log *.out* oscap_debug.log.*
+CLEANFILES = *.log *.out* oscap_debug.log.*
+
+TESTS_ENVIRONMENT = \
+	builddir=$(top_builddir) \
+	OSCAP_FULL_VALIDATION=1 \
+	$(top_builddir)/run
+
+# This is the entry script which should be responsible for running
+# all the tests from the 'foo' directory.
+TESTS = all.sh
+
+# This variable must contain all the files which are required to run tests.
+EXTRA_DIST = \
+	all.sh \
+	my_test_script.sh \
+	my_data.xml
+----
+
+[start=4]
+. Create an entry script, for example `all.sh`, inside the `tests/API/XCCDF/foo`
+   directory. This script should be responsible for executing all your test
+   scripts with specified arguments, for example:
+[[app-listing]]
+[source,bash]
+----
+#!/usr/bin/env bash
+
+set -o pipefail
+
+. ../../../test_common.sh
+
+function test_foo() {
+	...
+	return 0
+}
+
+test_init test_api_xccdf_foo.log
+
+test_run "Description 1" $srcdir/my_test_script.sh my_data.xml
+test_run "Description 2" test_foo
+
+test_exit
+----
+
+NOTE: The **entry script** is responsible for running all the tests from a
+specific directory and it is defined in the `TESTS` variable in the directory's
+`Makefile.am`.
+
+
+=== Common test library
+When writing tests for OpenSCAP you should use the common test library which is
+located at link:../../tests/test_common.sh.in[tests/test_common.sh.in].
+
+NOTE: The `configure` script will generate the `tests/test_common.sh` file from
+the link:../../tests/test_common.sh.in[tests/test_common.sh.in] adding
+configuration specific settings.
+
+You will need to source the `tests/test_common.sh` in your entry scripts to
+use the functions which it provides. When sourcing the file, you need to specify
+a relative path to it from the directory with your entry script. For example,
+when your entry script will be in the `tests/API/XCCDF/foo` directory:
+[[app-listing]]
+[source,bash]
+----
+#!/usr/bin/env bash
+
+set -o pipefail
+
+. ../../../test_common.sh
+
+test1
+test2
+...
+----
+
+
+==== Global variables exported by test library
+Always use `$OSCAP` variable instead of the plain `oscap` in your tests when
+calling the `oscap` command line tool. This is because tests might be run with
+`CUSTOM_OSCAP` variable.
+
+You can use `$XMLDIFF` in your tests which will call the
+link:../../tests/xmldiff.pl[tests/xmldiff.pl] script.
+
+It is also possible to do XPath queries using `$XPATH` variable, the usage is:
+[[app-listing]]
+[source,bash]
+----
+$XPATH FILE 'QUERY'
+----
+
+
+==== Setting shell attributes for your test scripts
+It is probably good to set `pipefail` option for all your test scripts so you
```

Comment:
it's not probably good, it's a good idea. don't make it sound weaker by adding probably/maybe/etc...